Add search and tag to metadata

marthacryan opened the pull request:
Using the code from https://github.com/jupytercalpoly/jupyterlab-code-snippets as a basis to add ability to search / filter by tags in metadata explorer. 

- [x] Add search bar and tag viewer
- [x] Add ability to search by name of metadata
- [x] Add ability to add a tag in the metadata editor
- [x] Add ability to filter by tags

This is the new feature:
![search](https://user-images.githubusercontent.com/6673460/96929565-ec223900-147f-11eb-90f1-3cec12b06ecc.gif)

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

marthacryan commented:
Found a bug: 
When searching, if a snippet is hidden because it doesn't match the search, when it reappears after erasing the search, the code area is blank. So for example, this is the code snippets before search "is":
![image](https://user-images.githubusercontent.com/6673460/96928577-62be3700-147e-11eb-8359-1eafa53fd2f8.png)

After removing search "is":
![image](https://user-images.githubusercontent.com/6673460/96928613-723d8000-147e-11eb-9025-09d42e395ffc.png)

marthacryan commented:
Fixed bug mentioned above - should be fully ready to play around with if anyone has time.

ptitzler commented:
Will test today

ptitzler commented:
- Is `cell tags` a good label? Not really sure what it means in this context

![image](https://user-images.githubusercontent.com/13068832/97043035-c865fe00-1526-11eb-9f58-0be0dc86787e.png)

- Perhaps we should auto-create tags for the snippet's language, since that is likely a commonly used filter criterion?

![image](https://user-images.githubusercontent.com/13068832/97043250-ffd4aa80-1526-11eb-89d3-89f5cc1a8ece.png)

ptitzler commented:
I'm running into an issue where a newly created code snippet isn't displayed in the list on the left hand side unless the screen is manually refreshed.

ptitzler commented:
What happened to drag-and-drop?

ptitzler commented:
Would be great if the list could be sorted alphabetically

![image](https://user-images.githubusercontent.com/13068832/97045032-e5e89700-1529-11eb-999f-68327d427c1c.png)

ptitzler commented:
![image](https://user-images.githubusercontent.com/13068832/97045598-dd449080-152a-11eb-8ed1-75b8b38ceebd.png)

- Unlike other pop-up messages, this one looks like a default alert window.
- The duplicate tag error message text should probably be a bit more formal, such as `A tag with this label already exists`.
- Perhaps we can just ignore the tag silently in this case, since it's already included in the set?

ptitzler commented:
Not sure tags should be case sensitive from a conceptual point of view. For example, what's the difference between "MARKDOWN" and "markdown" as far as the user is concerned?

![image](https://user-images.githubusercontent.com/13068832/97045924-7673a700-152b-11eb-807c-1e7eec5a1cfd.png)

ptitzler commented:
I'm not sure I understand the tag filter principle 

- Initially all (?) tags are displayed
![image](https://user-images.githubusercontent.com/13068832/97048156-d455be00-152e-11eb-85ee-a8f37c9b943e.png)


- Choose a tag - in this case only one snippet is matched (note that the `python` tag is no longer displayed)
![image](https://user-images.githubusercontent.com/13068832/97047736-3eba2e80-152e-11eb-94f7-61c461d32557.png)
- Choose a second tag from the list; now more snippets are shown and more tags are displayed
![image](https://user-images.githubusercontent.com/13068832/97047868-75904480-152e-11eb-9ced-cec254ee76a0.png)

ptitzler reviewed (changes requested):
See earlier review comments.

lresende commented on elyra/metadata/schemas/kfp.json:
```diff
@@ -94,6 +94,14 @@
             "category": "Cloud Object Storage"
           }
+        },
+        "tags": {
```
Why are we adding tags to runtime configuration?

lresende commented on elyra/metadata/schemas/runtime-image.json:
```diff
@@ -36,6 +36,14 @@
           "type": "string",
           "minLength": 1
+        },
+        "tags": {
```
same as above

lresende commented:
For the `filter by tag`, would this be ok assuming the user already clicked on the title and know the list is actually tags?

![image](https://user-images.githubusercontent.com/382917/97095757-68e81b00-1618-11eb-86a7-96e645f60552.png)

lresende commented on packages/metadata-common/src/Tags.tsx:
```diff
@@ -0,0 +1,249 @@
+
+    if (inputElement.value !== '' && event.keyCode === 13) {
+      if (this.state.tags.includes(inputElement.value)) {
+        alert('Duplicate Tag Name!');
```
we should use showDialog here.

lresende commented:
How, where we are populating the list of tags?

marthacryan commented:
Updates:
- [x] Sort tags alphabetically
- [x] Add searching for language in code snippet search
- [ ] Add `showDialog` (this one isn't currently working yet)
- [x] Remove "Cell Tags" header at the top of the list of tags
- [x] Changed list of tags to always display all tags

karlaspuldaro commented on packages/code-snippet/src/CodeSnippetWidget.tsx:
```diff
@@ -247,10 +251,52 @@ class CodeSnippetDisplay extends MetadataDisplay<ICodeSnippetDisplayProps> {
+    }
+
+    if (state.searchValue !== '' || state.filterTags.length !== 0) {
+      const newMetadata = props.metadata.filter((metadata: IMetadata) => {
+        return (
+          (state.searchValue !== '' &&
+            metadata.name.toLowerCase().includes(state.searchValue)) ||
+          metadata.display_name.toLowerCase().includes(state.searchValue) ||
+          metadata.metadata.language
+            .toLowerCase()
+            .includes(state.searchValue) ||
+          (metadata.metadata.tags &&
+            metadata.metadata.tags.some((tag: string) =>
+              state.filterTags.includes(tag)
+            ))
+        );
+      });
```
These checks won't cover all the edge cases, plus whenever I tried filtering it by a tag it didn't work (eg. I had only one code snippet tagged but all of them were still being displayed). 
![image](https://user-images.githubusercontent.com/25207344/97236354-493d2800-17bb-11eb-8b8b-48ebafca3a5a.png)

A suggested fix (tested and it works) and some refactoring could be:
```suggestion
      const filterTags = new Set(state.filterTags);
      const newMetadata = props.metadata.filter((metadata: IMetadata) => {
        const searchValue = state.searchValue.toLowerCase().trim();
        return searchValue
            && (metadata.name.toLowerCase().includes(searchValue) 
                || metadata.display_name.toLowerCase().includes(searchValue) 
                || metadata.metadata.language.toLowerCase().includes(searchValue)) 
            || metadata.metadata.tags 
                && metadata.metadata.tags.some((tag: string) => filterTags.has(tag));
      });
```

karlaspuldaro reviewed (changes requested):
A few comments with tested changes proposed.
Also agree with @ptitzler point on case sensitivity, I think lowercasing the tag values on display would make more sense and avoid duplicates.

marthacryan commented on packages/metadata-common/src/Tags.tsx:
```diff
@@ -0,0 +1,259 @@
+
+    if (inputElement.value !== '' && event.keyCode === 13) {
+      if (this.state.tags.includes(inputElement.value)) {
+        await showDialog({
```
Wow thank you for figuring that out!!

karlaspuldaro commented on packages/metadata-common/src/Tags.tsx:
```diff
@@ -0,0 +1,260 @@
+          title: 'Duplicate Tag Name!',
+          buttons: [Dialog.okButton()]
+        }).then((value: any) => {
+          console.log('VALUE');
```
we can remove this statement

karlaspuldaro commented:
When we have too many tags the Save & Close button stays in the way
![image](https://user-images.githubusercontent.com/25207344/97333805-5527f900-1852-11eb-9b9c-df36897bcdf9.png)
To make the button still visible in the editor but away from the tags maybe it could be moved to the bottom right?

marthacryan commented:
Fixed the css issues:
![image](https://user-images.githubusercontent.com/6673460/97345941-c10d6000-1858-11eb-96ce-69377f5bd24b.png)

ptitzler commented:
I ran another quick test using the updated version. Creating a new code snippet the user is currently presented with something like this:

![image](https://user-images.githubusercontent.com/13068832/97345008-c792dc00-1846-11eb-90b8-313d3e278a45.png)

To make it easier for the user, we should automatically assign language tags to code snippets. For example:
- User creates a new code snippet choosing `Java` as language (but no tags)
- User saves the new code snippet.
- The code snippet panel is refreshed and `java` is displayed as one of the tags.
- The code snippet is automatically associated with this tag.

Benefit:
 - when editing a code snippet the user isn't presented with language tags that don't apply (less screen real estate is used)
 - the user doesn't have to explicitly choose the language specific tag (fewer steps are required)

ptitzler commented:
Currently the free form search expression and the selected tags are ORed, to compute the results.  From a user's point of view, wouldn't it be more likely that the user wants to see those code snippets that meet both conditions, which in essence represents a faceted search that many search engines are using.

- current behavior: show me the code snippets that contain the string MIT or were tagged with `one more tag`
- desirable (?) behavior: show me the code snippets that contain the string MIT in the name and were tagged with `one more tag`

![image](https://user-images.githubusercontent.com/13068832/97346926-50127c00-1849-11eb-8a8d-5d2446b2b466.png)

marthacryan commented:
Just pushed a change to the searching - @ptitzler @karlaspuldaro feel free to try it out! And @karlaspuldaro let me know if you think I should refactor the code a little

marthacryan commented on packages/code-snippet/src/CodeSnippetWidget.tsx:
```diff
@@ -247,10 +251,55 @@ class CodeSnippetDisplay extends MetadataDisplay<ICodeSnippetDisplayProps> {
     );
   }
 
+  static getDerivedStateFromProps(
```
Sounds good - I was thinking that too

marthacryan replied:
Added a method called `matchesSearch` that returns a boolean if it matches the tag / search.

marthacryan commented:
@ptitzler Would you mind making a separate issue for the language tag stuff? We can add that as a separate PR

kevin-bates reviewed (commented):
Thanks Martha - just had a few comments.

karlaspuldaro commented on packages/metadata-common/src/MetadataWidget.tsx:
```diff
@@ -152,13 +179,111 @@ export class MetadataDisplay<
+    return tags;
+  }
+
+  static matchesSearch(
```
I agree, the one in CodeSnippetsWidget is not being used anymore

marthacryan replied:
The difference is that the `language` field is checked within the `CodeSnippetsWidget` implementation

marthacryan commented on packages/code-snippet/src/CodeSnippetWidget.tsx:
```diff
@@ -247,10 +251,36 @@ class CodeSnippetDisplay extends MetadataDisplay<ICodeSnippetDisplayProps> {
+    // True if search string is in name, display_name, or language of snippet
+    // or if the search string is empty
+    const matchesSearch =
+      metadata.name.toLowerCase().includes(searchValue) ||
```
I added the extra method `matchesSearch` so that if a child class of `MetadataWidget` (in this case, `CodeSnippetWidget`) wanted to match extra criteria in the search (in this case, the `language` field) they could override the default (which just matches the `name` or the `display_name`.) I'm assuming that the child class won't override the method that calls `matchesSearch` (`getDerivedStateFromProps`) and including the checks for things like empty strings and making the `searchValue` lowercase so that when a child class implements `matchesSearch` it won't have to use duplicate code. Does that make sense?

kevin-bates replied:
I see.  I didn;t realize these were subclasses.  If that's the case, shouldn't this implementation call super.matchesSearch and, if no match, then check `language`?  However, because of the 'anded' behavior with tags, it seems like the two forms should be split (if you want to support subclassing like this).

I would suggest something like the following:
In the base class, implement `matchesSearch()` or whatever the primary name is.  This implementation calls two methods, one that matches via the metadata fields, let's say, `matchesMetadata()` and the other that matches tags, let's say, `matchesTags()`.  So you will have this in the base class (using python syntax):

```python
def matchesSearch(...):
    return matchesMetadata() and matchesTags()
```
Then, CodeSnippets would override `matchesMetadata()` with something like the following...

```python
def matchesMetadata():
    matches = super.matchesMetadata()
    if not matches:
        matches = <perform matching logic again `language` and any other code-snippet-only fields>
    return matches
```
Since the implementation for the tags matching is essentially metadata-type-independent, it will likely rarely be overridden.  LIkewise, the primary method, `matchesSearch()` would also rarely be overridden.

ajbozarth replied:
as I discussed with @marthacryan , this is a great idea. Having separate `matchesSearch` and `matchesTags` will allow for better extensibility. Similarly about calling `super` rather than reimplement that functionality

ajbozarth replied:
though it's called in the function that calls this, I feel that `toLower` should still be called  on `searchValue` here. A function should not make assumptions about what other functions that call it do with the params (unless absolutely necessary)

lresende reviewed (approved):
Other then https://github.com/elyra-ai/elyra/issues/1019 it LGTM

ptitzler commented:
I noticed a bit of an undesirable side effect in the runtime images wizard, since it currently uses the default JSON dump to "visualize" the metadata.

![image](https://user-images.githubusercontent.com/13068832/97625457-c2b85e80-19e5-11eb-9cc5-22ede225c197.png)

ajbozarth commented on packages/metadata-common/src/MetadataWidget.tsx:
```diff
@@ -152,13 +179,111 @@ export class MetadataDisplay<
+    return matchesSearch && matchesTags;
+  }
+
+  static getDerivedStateFromProps(
```
this can't be `static` if we want to override the `matchesSearch` call found in this function in child classes

ajbozarth reviewed (changes requested):
Leaving this review here while I continue a deeper review. I've already discussed all these points with @marthacryan in a Webex call.

marthacryan commented on packages/metadata-common/src/MetadataWidget.tsx:
```diff
@@ -152,13 +179,111 @@ export class MetadataDisplay<
     );
   }
 
+  filteredMetadata = (searchValue: string, filterTags: string[]): void => {
```
I tried changing this but it actually broke the functionality, I think because it's passed as a value to `FilterTools`.

marthacryan commented:
@ajbozarth @kevin-bates I tried switching in the non-static (but deprecated) `componentWillUpdate` but it doesn't seem to work in the same way that `getDerivedStateFromProps` does (the state isn't being updated). I also think that we should avoid using a deprecated method if we don't have to. I think given the inheritance issues with static methods, we should just go back to the original code that overrides `getDerivedStateFromProps` in `CodeSnippetDisplay` and then refactor later on (we just had a discussion with Nic and he suggested a larger refactor of the inheritance structure - I plan to create an issue to describe it further). It's a little bloated to duplicate all of the code, but it's also the safest way of doing it for now. Does that sound okay?

karlaspuldaro commented on packages/metadata-common/src/MetadataWidget.tsx:
```diff
@@ -152,13 +185,113 @@ export class MetadataDisplay<
+    filterTags: Set<string>,
+    metadata: IMetadata
+  ): boolean {
+    // True if search string is in name, display_name, or language of snippet
```
This comment can be updated as language is not being filtered in this method

ajbozarth commented on packages/metadata-common/src/MetadataWidget.tsx:
```diff
@@ -129,7 +150,13 @@ export class MetadataDisplay<
+        key={metadata.name}
+        className={METADATA_ITEM}
+        style={{
+          display: this.state.metadata.includes(metadata) ? 'flex' : 'none'
```
this was fixed in code snippet but not here in metadata widget

ajbozarth commented on packages/metadata-common/src/MetadataWidget.tsx:
```diff
@@ -152,13 +185,113 @@ export class MetadataDisplay<
+      metadata.display_name.toLowerCase().includes(searchValue);
+    // True if there are no tags selected or if there are tags that match
+    // tags of metadata
+    const matchesTags =
```
I believe we should separate `matchesTags` into it's own function that is called separately by `getDerivedStateFromProps`

ajbozarth commented on packages/metadata-common/src/MetadataWidget.tsx:
```diff
@@ -152,13 +179,111 @@ export class MetadataDisplay<
+    props: IMetadataDisplayProps,
+    state: IMetadataDisplayState
+  ): IMetadataDisplayState {
+    if (state.searchValue === '' && state.filterTags.length === 0) {
```
Echoing @kevin-bates comment, why are these if statements needed, it seems like we could remove this if block and the if wrapping around the next section and get the same results

marthacryan replied:
Not sure why but when you remove this block it doesn't initially show any of the metadata instances - it only shows metadata when you search / filter by tags. Should I look into this more?

ajbozarth commented on packages/metadata-common/src/MetadataWidget.tsx:
```diff
@@ -129,7 +156,13 @@ export class MetadataDisplay<
   // Render display of metadata list
```
As mentioned in a previous comment by @ptitzler we should update `renderExpandableContent` to ignore the `tags` section of the json. This can be handled in a follow up PR if desired though

ajbozarth reviewed (commented):
I have not yet reviewed the new `FilterTools` or `Tags` files, and have not played with the features locally yet, but here's a handful of review comments on current changes

ajbozarth commented on packages/metadata-common/src/Tags.tsx:
```diff
@@ -0,0 +1,258 @@
+const METADATA_EDITOR_TAG_PLUS_ICON = 'elyra-editor-tag-plusIcon';
+const METADATA_EDITOR_TAG_LIST = 'elyra-editor-tagList';
+
+export class MetadataEditorTags extends React.Component<
```
Is there a reason this file is called `Tags` when this class is called `MetadataEditorTags` given this is a single class file the two should match

ajbozarth commented on packages/metadata-common/src/Tags.tsx:
```diff
@@ -0,0 +1,258 @@
+    this.props.handleChange(this.state.selectedTags, this.state.tags);
+  }
+
+  handleClickHelper(
```
This function name is kinda vague, it seems to handle updating css on click. iiuc perhaps `updateTagsCss` might be a better name (unless I'm misunderstanding what this is doing)

ajbozarth commented on packages/metadata-common/src/FilterTools.tsx:
```diff
@@ -0,0 +1,207 @@
+    );
+  }
+
+  handleClickHelper(
```
same notes as other `handleClickHelper`

ajbozarth reviewed (commented):
A few comments on the new files, otherwise no big issues with them

ajbozarth commented:
UI/UX review comments:

- Clicking the icon in a tag, the check mark or plus, doesn't "click" the tag as I would expect it to (in fact only the text in the tag is clickable)

- Odd css happening with the down arrow icon:
<img width="165" alt="Screen Shot 2020-11-03 at 12 20 42 PM" src="https://user-images.githubusercontent.com/13952758/98036684-510a5700-1dcf-11eb-925c-2f75f05a8aae.png">
might be caused by using css instead of LabIcon

- If you filter on tags then close the filter on tags section there is no way to determine that it's still filtering on tags, this may confuse users who filter then leave the snippet UI and come back a long time later

marthacryan commented:
![image](https://user-images.githubusercontent.com/6673460/98299745-ca47ac80-1f7d-11eb-8633-20c129826ef3.png)
Updated the "Filter by Tags" area to use a labicon of the caret up icon rather than the css it was using before. I added the behavior that when the "Filter by Tags" area is closed, it doesn't filter by any tags. I also changed the button to be completely "clickable", and not just the text. 

It looks good except there's some weird css stuff going on in safari due to these changes - @ajbozarth maybe you could check it out? I'm not really finding anything to fix it (see below)
![image](https://user-images.githubusercontent.com/6673460/98305139-eb60cb00-1f86-11eb-8b9e-1ed3de52d6a3.png)

ajbozarth commented:
I've pushed a fix for the safari css you found.

As for the "carrot" icon under Filter By Tags, I think we should move it next to the text and have it swap between a up and down carrot depending on if the tags display is open, making it more clear that its a dropdown. Also I would suggest we use the carrot icons already in lab rather than adding our own.

As for the disabling of the filter when it's closed, I think a better implementation might be to let the user close it and still be filtered, but to highlight the Filter By Tags text to indicate whether filtering is being applied.

ajbozarth commented:
while testing my above fix I noticed some really odd behavior with the tags that should be further looked into as well, including a screen capture gif below to show it
![Screen Recording 2020-11-05 at 4 04 11 PM](https://user-images.githubusercontent.com/13952758/98310304-72ab4000-1f81-11eb-80ae-cfba570f8ce8.gif)

marthacryan commented:
If we're changing the design of the "filter by tags" button, what are your thoughts on having a tag button like the ToC uses? (see below)
![image](https://user-images.githubusercontent.com/6673460/98390559-27457000-201b-11eb-9b2c-9cd18769d35a.png)

marthacryan commented:
Here's the updated filter button:
![tags](https://user-images.githubusercontent.com/6673460/98425571-3b10c680-205b-11eb-8f88-99f12880a557.gif)

marthacryan commented:
Small update to the size of the gray box that appears on hover (makes it tighter around the icon):
![image](https://user-images.githubusercontent.com/6673460/98428434-13276000-2067-11eb-9e64-10f35ff2aa4a.png)

marthacryan commented:
Fixed safari bug.

marthacryan commented:
![image](https://user-images.githubusercontent.com/6673460/98429547-e0806600-206c-11eb-88d3-47a2b667ec5f.png)
Thoughts on this indicator for when filtering by tags and the tag area is closed?

ajbozarth reviewed (approved):
LGTM

lresende reviewed (approved):
LGTM
